Fixes for HostIPC tests to work when Docker has SELinux support enabled.

filbranden opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

Fixes for HostIPC tests to work when Docker has SELinux support enabled.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

N/A

**Special notes for your reviewer**:

The core of the matter is to use `ipcs` from util-linux rather than the one from busybox. The typical SELinux policy has enough to allow Docker containers (running under svirt_lxc_net_t SELinux type) to access IPC information by reading the contents of the files under /proc/sysvipc/, but not by using the shmctl etc. syscalls.

The `ipcs` implementation in busybox will use `shmctl(0, SHM_INFO, ...)` to detect whether it can read IPC info (see source code [here](https://git.busybox.net/busybox/tree/util-linux/ipcs.c?h=1_28_0#n138)), while the one in util-linux will prefer to read from the /proc files directly if they are available (see source code [here](https://github.com/karelzak/util-linux/blob/v2.27.1/sys-utils/ipcutils.c#L108)).

It turns out the SELinux policy doesn't allow the shmctl syscalls in an unprivileged container, while access to it through the /proc interface is fine. (One could argue this is a bug in the SELinux policy, but getting it fixed on stable OSs is hard, and it's not that hard for us to test it with an util-linux `ipcs`, so I propose we do so.)

This PR also contains a refactor of the code setting IpcMode, since setting it in the "common options" function is misleading, as on containers other than the sandbox, it ends up always getting overwritten, so let's only set it to "host" in the Sandbox.

It also has a minor fix for the `ipcmk` call, since support for size suffix was only introduced in recent versions of it.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

filbranden committed:
Don't assume ipcmk command supports size suffix.

Expand the use of "1M" to the corresponding number of bytes, since
support for size suffix was only added to `ipcmk` in util-linux 2.27
which is not yet available in some Linux distributions.

Tested by running `make test-e2e-node` against distributions with ipcmk
that supports and doesn't support the suffix syntax, all of them passed.

filbranden commented:
For the record, the SELinux fix for this can be done using the following module (save it to a file named `docker_ipc.te`):

```selinux
module docker_ipc 1.0;

require {
        type unconfined_t;
        type kernel_t;
        type svirt_lxc_net_t;
        class shm { associate getattr unix_read };
        class system ipc_info;
}

#============= svirt_lxc_net_t ==============

allow svirt_lxc_net_t kernel_t:system ipc_info;
allow svirt_lxc_net_t unconfined_t:shm { associate getattr unix_read };
```

Compile and install it with:

```shell
$ checkmodule -M -m -o docker_ipc.mod docker_ipc.te
$ semodule_package -o docker_ipc.pp -m docker_ipc.mod
$ sudo semodule -i docker_ipc.pp
```

After that SELinux module is installed, `ipcs -m` will work inside a container that has IpcMode set to "container:\<id>".

Random-Liu commented:
/ok-to-test

Random-Liu commented:
> The net result is more or less the same (since with the old approach they would essentially get "host" through the Pod sandbox container), but doing it explicitly this way has SELinux implications.

What selinux implication is that? I think it's good to know. AFAIK, both cri-containerd and dockershim have the same implementation - sandbox container has host namespace, other containers share sandbox container's namespace.

If this is a problem on SELinux, we may need to understand why, and whether it makes sense to make this change. And if it makes sense, I'll make the same change in cri-containerd. :)

filbranden commented:
> And why only IPC namespace sharing has this problem?

The problem is actually with running `ipcs` command inside a container. By default, the SELinux policy prevents that from successfully working.

By default, this is what happens:

```
$ docker run busybox ipcs -m
kernel not configured for shared memory
```

So setting "--ipc host" in Docker makes the container run in a different SELinux context, which lets it run ipcs successfully:

```
$ docker run --ipc host busybox ipcs -m
------ Shared Memory Segments --------
key        shmid      owner      perms      bytes      nattch     status
```

Here's the difference in SELinux context with and without `--ipc host`:

```
$ docker run ubuntu ps -Z
LABEL                             PID TTY          TIME CMD
system_u:system_r:svirt_lxc_net_t:s0:c468,c779 1 ? 00:00:00 ps
$ docker run --ipc host ubuntu ps -Z
LABEL                             PID TTY          TIME CMD
system_u:system_r:spc_t:s0          1 ?        00:00:00 ps
```

Setting `--ipc` to point to another container doesn't trigger that SELinux context, so the `ipcs` command keeps failing.

See also my previous comment in this PR, there's some SELinux policy module that could be used to allow this.

NOTE: While testing this, I also noticed that this problem exists in busybox's `ipcs` only, it doesn't seem to affect the `ipcs` in e.g. Ubuntu. It's probably because the `ipcs` in busybox is doing additional work that Ubuntu's one is not doing.

> If this is a problem on SELinux, we may want to understand why, and whether it makes sense to make this change.

So I'm thinking this doesn't really make sense... I'm leaning towards:

1. Changing `modifyCommonNamespaceOptions`, so it doesn't set `hostConfig.IpcMode = namespaceModeHost` and move that to `modifySandboxNamespaceOptions`, since it's only relevant for the sandbox. That would clear up the confusion about it being overwritten later.

2. Either disable this test when SELinux is enabled, or maybe test it using a different container, such as "ubuntu" rather than "busybox". Do you think this would be acceptable?

Cheers!
Filipe

filbranden committed:
Add new e2e-test container to export ipcs from util-linux

This container will be used to exercise the HostIPC functionality in
e2e-node tests.

The version of `ipcs` shipped in busybox performs operations that get
blocked by SELinux on hosts where it is enabled. The version of `ipcs`
in util-linux does not perform those operations, rather it checks
whether the /proc files it needs are available and proceeds to reading
from them directly.

Using `ipcs` from util-linux makes these tests pass, even when running
under SELinux enabled, so let's use them here.

Tested: On a host where Docker with SELinux enabled:

- Checked that `ipcs` from busybox does not work:

  $ docker run busybox ipcs -m
  kernel not configured for shared memory

- Checked that the one from this container does work:

  $ docker run gcr.io/kubernetes-e2e-test-images/ipc-utils-amd64:1.0 ipcs -m
  ------ Shared Memory Segments --------
  key        shmid      owner      perms      bytes      nattch     status

filbranden committed:
Refactor handling of IpcMode for the actual container

Having the field set in modifyCommonNamespaceOptions is misleading,
since for the actual container it is later unconditionally overwritten
to point to the sandbox container.

So let's move its setting to modifyHostOptionsForSandbox (renamed from
modifyHostNetworkOptionForSandbox as it's not about network only), since
that reflects what actually happens in practice.

This commit is purely a refactor, it doesn't change any behavior.

filbranden committed:
Use ipc-utils container in HostIPC tests.

This ensures the `ipcs` command from util-linux will be used, which
succeeds when Docker is running with SELinux enabled (while the one from
busybox fails.)

Tested: On a host with Docker running with SELinux enabled:

  $ make test-e2e-node REMOTE=true FOCUS="host IPC"

  • [SLOW TEST:17.272 seconds] (passed)
  [k8s.io] Security Context
    when creating a pod in the host IPC namespace
      should show the shared memory ID in the host IPC containers

  • [SLOW TEST:20.419 seconds] (passed)
  [k8s.io] Security Context
    when creating a pod in the host IPC namespace
      should not show the shared memory ID in the non-hostIPC containers

  Ran 2 of 257 Specs in 43.934 seconds
  SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 255 Skipped

filbranden commented:
@Random-Liu Please take a look. I also updated the PR description to reflect what we're doing here. I already pushed the `ipc-utils` container defined here to gcr.io (hope that's OK!)

Cheers,
Filipe

filbranden commented:
Looks like a flake... so:

/test pull-kubernetes-e2e-kops-aws

Random-Liu commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -143,18 +145,17 @@ func modifyContainerNamespaceOptions(nsOpts *runtimeapi.NamespaceOption, podSand
 
 // modifyHostNetworkOptionForSandbox applies NetworkMode/UTSMode to sandbox's dockercontainer.HostConfig.
-func modifyHostNetworkOptionForSandbox(hostNetwork bool, network *knetwork.PluginManager, hc *dockercontainer.HostConfig) {
+func modifyHostNetworkOptionForSandbox(hostNetwork bool, hostIpc bool, network *knetwork.PluginManager, hc *dockercontainer.HostConfig) {
```
Rename this function?

filbranden replied:
Do you have any suggestions? Perhaps just `modifyHostOptionsForSandbox` (note "Options" plural.)

A few lines below, there's `modifyHostNetworkOptionForContainer` which also sets IpcMode and UTSMode, so I'm thinking for consistency we should rename that one too... Agreed?

Random-Liu replied:
I prefer `modifyHostOptionsForSandbox`. :) Since you've passed in `hostIpc` here.

filbranden replied:
Done! Also renamed modifyHostNetworkOptionForContainer -> modifyHostOptionsForContainer for consistency (that one also sets hostIpc, even though it doesn't take it as an argument...) PTAL.

Random-Liu commented:
LGTM with one nit.

filbranden reviewed (commented):
Thanks for the review! I have a small follow up question.

filbranden commented:
Let's try this again:

/test pull-kubernetes-e2e-kops-aws

Random-Liu commented:
/lgtm
/approve no-issue

filbranden commented:
/assign @vishh 

Vish, can you please approve for OWNERS?

Thanks!
Filipe

vishh commented:
/approve

filbranden commented:
Retry...

/test pull-kubernetes-unit

ixdy commented:
This PR appears to have broken the crossbuild:
```
I0131 05:29:12.462] # k8s.io/kubernetes/pkg/kubelet/dockershim
I0131 05:29:12.462] pkg/kubelet/dockershim/helpers_windows.go:52:3: undefined: modifyHostNetworkOptionForContainer
```

cblecker commented:
I opened #59121 for the cross build failure
